github-extension/background.js

```diff
@@ -0,0 +1,113 @@
+    const req = new XMLHttpRequest();
+    const apiUrl = `${BASE_URL}?access_token=${API_KEY}`;
+    req.addEventListener("load", () => {
+      const priorityCards = [];
+      const fillerCards = [];
+      const {data} = JSON.parse(req.response);
+      const {viewer} = data;
+      const {login, contributedRepositories} = viewer;
+      for (const {node: repo} of contributedRepositories.edges) {
+        const {nameWithOwner, pullRequests, issues} = repo;
+        for (const {node: pr} of pullRequests.edges) {
+          const {author, number, title, assignees, url, reviewRequests, updatedAt} = pr;
+          const isAuthor = author.login === login;
+          const isAssigned = assignees.edges.some(({node: assignee}) => assignee.login === login);
+          const isReviewRequested = reviewRequests.edges.some(({node: rr}) => rr.reviewer.login === login);
+          if (isAuthor || isAssigned || isReviewRequested) {
+            let significance;
+            if (isReviewRequested) {
+              significance = "Your review requested";
+            } else if (isAssigned) {
+              significance = "Assigned to you";
+            } else {
+              significance = "You created";
+            }
+            priorityCards.push({
+              title: `PR #${number}: ${significance}`,
+              description: title,
+              url,
+              image: author.avatarUrl,
+              hostname: nameWithOwner,
+              updatedAt
+            });
+          }
+        }
+        if (issues) {
+          for (const {node: issue} of issues.edges) {
+            const {author, number, title, assignees, url, updatedAt} = issue;
+            const isAuthor = author.login === login;
+            const isAssigned = assignees.edges.some(({node: assignee}) => assignee.login === login);
+            if (isAuthor || isAssigned) {
+              let significance;
+              if (isAssigned) {
+                significance = "Assigned to you";
+              } else {
+                significance = "You reported";
+              }
+              priorityCards.push({
+                title: `Issue #${number}: ${significance}`,
+                description: title,
+                url,
+                image: author.avatarUrl,
+                hostname: nameWithOwner
+              });
+            } else {
+              fillerCards.push({
+                title: `Issue #${number}`,
+                description: title,
+                url,
+                image: author.avatarUrl,
+                hostname: nameWithOwner,
+                updatedAt
+              });
+            }
+          }
+        }
+      }
+      priorityCards.sort((a, b) => Date.parse(b.updatedAt) - Date.parse(a.updatedAt));
+      fillerCards.sort((a, b) => Date.parse(b.updatedAt) - Date.parse(a.updatedAt));
+      const allCards = priorityCards.concat(fillerCards);
+      allCards.forEach(card => delete card.updatedAt);
+      resolve(allCards);
```

Comment:
If somehow we don't get anything returned in the body, will this fail? Should we check that the response has been successful and also returned an non-empty data, else just resolve on an empty array?

Or does this not really matter since this is a fake example?

Reply:
I don't think it would fail as priorityCards and fillerCards are both initialised at empty arrays and if nothing gets pushed to them then they'll still be empty arrays. The sorting will have no effect, the concat will return an empty array, and the allCards.forEach won't be called, so I think it would just resolve with an empty array